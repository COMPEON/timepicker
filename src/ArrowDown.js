import React from 'react'
import styled from 'styled-components'

import theme from './defaultTheme'

const ArrowDown = props => (
  <svg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' {...props}>
    <path d='M7.41,8.59L12,13.17l4.59-4.58L18,10l-6,6l-6-6L7.41,8.59z'/>
    <path fill='none' d='M0,0h24v24H0V0z'/>
  </svg>
)

const StyledArrowDown = styled(ArrowDown)`
  cursor: pointer;
  fill: ${props => props.theme.primaryColor};
`

StyledArrowDown.defaultProps = { theme }

export default StyledArrowDown

