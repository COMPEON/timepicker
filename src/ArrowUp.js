import React from 'react'
import styled from 'styled-components'

import theme from './defaultTheme'

const ArrowUp = props => (
  <svg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' {...props}>
    <path d='M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z' />
    <path d='M0 0h24v24H0z' fill='none' />
  </svg>
)

const StyledArrowUp = styled(ArrowUp)`
  cursor: pointer;
  fill: ${props => props.theme.primaryColor};
`

StyledArrowUp.defaultProps = { theme }

export default StyledArrowUp
