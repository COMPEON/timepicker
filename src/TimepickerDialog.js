import React from 'react'
import styled from 'styled-components'

import ArrowDown from './ArrowDown'
import ArrowUp from './ArrowUp'
import { formatNumber } from './utils'
import theme from './defaultTheme'

const TimepickerDialog = styled.div`
  align-items: center;
  background: white;
  border-radius: 5px;
  border: 1px solid ${props => props.theme.primaryColor};
  box-shadow: 0 1px 3px #d3d3d380, 0 1px 3px #d3d3d380;
  color: ${props => props.theme.primaryColor};
  display: flex;
  font-size: 22px;
  justify-content: center;
  position: absolute;
  user-select: none;
  width: 190px;
  z-index: 999;
`

TimepickerDialog.defaultProps = { theme }

const TimepickerPart = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  margin: auto 12px;
`

export const TimepickerBlock = ({ number, onDownClick, onUpClick }: TTimepickerBlockProps) => (
  <TimepickerPart>
    <ArrowUp onClick={onUpClick} size={36} />
    {formatNumber(number)}
    <ArrowDown onClick={onDownClick} size={36} />
  </TimepickerPart>
)

export default TimepickerDialog
