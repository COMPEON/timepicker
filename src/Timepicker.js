import * as React from 'react'
import PropTypes from 'prop-types'

import TimepickerDialog, { TimepickerBlock } from './TimepickerDialog'
import { addOrSubtractMinutes, parseTime } from './utils'
import theme from './defaultTheme'

class Timepicker extends React.Component {
  static defaultProps = {
    color: '#27718c',
    minutesPerStep: 15,
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    value: '00:00'
  }

  static propTypes = {
    className: PropTypes.string,
    color: PropTypes.string,
    dialogClassName: PropTypes.string,
    from: PropTypes.string,
    innerRef: PropTypes.func,
    minutesPerStep: PropTypes.number,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    to: PropTypes.string,
    value: PropTypes.string.isRequired
  }

  state = {
    dialogOpen: false
  }

  componentDidMount () {
    this.addDOMEvents()
    this.ensureValueInRange()
  }

  componentWillUnmount () {
    this.removeDOMEvents()
  }

  addDOMEvents = () => {
    document.addEventListener('click', this.handleDocumentClick)
    document.addEventListener('keydown', this.handleDocumentKeyPress)
    if (this.inputElement) this.inputElement.addEventListener('keydown', this.handleDocumentKeyPress)
  }

  removeDOMEvents = () => {
    document.removeEventListener('click', this.handleDocumentClick)
    document.removeEventListener('keydown', this.handleDocumentKeyPress)
    if (this.inputElement) this.inputElement.removeEventListener('keydown', this.handleDocumentKeyPress)
  }

  ensureValueInRange = () => this.handleMinutesChange(0)

  closeDialog = () => {
    if (!this.state.dialogOpen) return

    const nextValue = this.ensureValueInRange()

    this.setState({ dialogOpen: false })
    this.props.onBlur(nextValue)
  }

  openDialog = () => {
    if (this.state.dialogOpen) return

    this.setState({ dialogOpen: true })
  }

  handleDocumentKeyPress = event => {
    if (event.key !== 'Tab') return

    this.closeDialog()
  }

  handleDocumentClick = event => {
    const element = event.target

    if (!this.dialogElement || !this.inputElement || !(element instanceof Node)) return
    if (this.dialogElement.contains(element) || this.inputElement.contains(element)) return

    this.closeDialog()
  }

  handleDialogRef = element => {
    this.dialogElement = element
  }

  handleInputRef = (element, ...args) => {
    this.inputElement = element

    if (this.props.innerRef) {
      this.props.innerRef.call(null, element, ...args)
    }
  }

  handleChange = event => this.props.onChange(event.target.value || '')

  handleFocus = event => {
    this.props.onFocus(event)
    this.openDialog()
  }

  handleHoursChange = delta => this.handleMinutesChange(60 * delta)

  handleMinutesChange = delta => {
    const { from, onChange, to, value } = this.props
    const nextValue = addOrSubtractMinutes(value, delta, { from, to })

    if (nextValue !== value) onChange(nextValue)

    return nextValue
  }

  handleWheel = event => {
    const { minutesPerStep } = this.props
    const delta = event.deltaY > 0 ? +minutesPerStep : -minutesPerStep

    event.preventDefault()

    this.handleMinutesChange(delta)
  }

  render () {
    const { dialogOpen } = this.state
    const {
      children,
      className,
      dialogClassName,
      inputClassName,
      minutesPerStep,
      value
    } = this.props
    const { hours, minutes } = parseTime(value)

    return (
      <div className={className}>
        {React.cloneElement(
          React.Children.only(children), {
            className: inputClassName,
            ref: this.handleInputRef,
            onBlur: () => {},
            onChange: this.handleChange,
            onFocus: this.handleFocus,
            onTouchMove: this.handleWheel,
            onWheel: this.handleWheel,
            value
          })
        }
        {dialogOpen && (
          <TimepickerDialog
            className={dialogClassName}
            innerRef={this.handleDialogRef}
            onTouchMove={this.handleWheel}
            onWheel={this.handleWheel}
          >
            <TimepickerBlock
              number={hours}
              onDownClick={() => this.handleHoursChange(-1)}
              onUpClick={() => this.handleHoursChange(1)}
            />
            :
            <TimepickerBlock
              number={minutes}
              onDownClick={() => this.handleMinutesChange(-minutesPerStep)}
              onUpClick={() => this.handleMinutesChange(+minutesPerStep)}
            />
          </TimepickerDialog>
        )}
      </div>
    )
  }
}

export default Timepicker
