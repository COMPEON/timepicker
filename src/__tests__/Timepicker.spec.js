import React from 'react'
import { mount, shallow } from 'enzyme'

import Timepicker from '../Timepicker'
import TimepickerDialog from '../TimepickerDialog'
import ArrowDown from '../ArrowDown'
import ArrowUp from '../ArrowUp'

describe('<Timepicker />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Timepicker>
        <input />
      </Timepicker>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('opens the dialog when focused', () => {
    const wrapper = shallow(
      <Timepicker>
        <input />
      </Timepicker>
    )

    wrapper.find('input').simulate('focus')

    expect(wrapper).toMatchSnapshot()
  })

  it('properly delegates different class names', () => {
    const wrapper = mount(
      <Timepicker
        className='className'
        dialogClassName='dialogClassName'
        inputClassName='inputClassName'
      >
        <input />
      </Timepicker>
    )

    wrapper.find('input').simulate('focus')

    expect(wrapper.find('div').at(0).props().className).toEqual('className')
    expect(wrapper.find(TimepickerDialog).props().className).toEqual('dialogClassName')
    expect(wrapper.find('input').props().className).toEqual('inputClassName')
  })

  it('changes the time on a mousewheel event', () => {
    const changeHandler = jest.fn()
    const wrapper = mount(
      <Timepicker onChange={changeHandler}>
        <input />
      </Timepicker>
    )

    wrapper.find('input').simulate('focus')
    wrapper.find(TimepickerDialog).simulate('wheel')

    expect(changeHandler).toHaveBeenCalled()
  })

  it('changes the time when arrows are clicked', () => {
    const changeHandler = jest.fn()
    const wrapper = mount(
      <Timepicker
        onChange={changeHandler}
        minutesPerStep={15}
        value='20:00'
      >
        <input />
      </Timepicker>
    )

    wrapper.find('input').simulate('focus')

    const downArrows = wrapper.find(ArrowDown)
    const upArrows = wrapper.find(ArrowUp)

    expect(downArrows.length).toBe(2)
    expect(upArrows.length).toBe(2)

    downArrows.at(0).simulate('click')
    downArrows.at(1).simulate('click')

    expect(changeHandler).toHaveBeenCalledTimes(2)
    expect(changeHandler).toHaveBeenCalledWith('19:00')
    expect(changeHandler).toHaveBeenCalledWith('19:45')

    upArrows.at(0).simulate('click')
    upArrows.at(1).simulate('click')

    expect(changeHandler).toHaveBeenCalledTimes(4)
    expect(changeHandler).toHaveBeenCalledWith('21:00')
    expect(changeHandler).toHaveBeenCalledWith('20:15')
  })
})
