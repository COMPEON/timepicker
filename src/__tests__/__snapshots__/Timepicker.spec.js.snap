// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timepicker /> opens the dialog when focused 1`] = `
<div>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onTouchMove={[Function]}
    onWheel={[Function]}
    value="00:00"
  />
  <styled.div
    innerRef={[Function]}
    onTouchMove={[Function]}
    onWheel={[Function]}
    theme={
      {
        "primaryColor": "#27718c",
      }
    }
  >
    <TimepickerBlock
      number={0}
      onDownClick={[Function]}
      onUpClick={[Function]}
    />
    :
    <TimepickerBlock
      number={0}
      onDownClick={[Function]}
      onUpClick={[Function]}
    />
  </styled.div>
</div>
`;

exports[`<Timepicker /> renders correctly 1`] = `
<div>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onTouchMove={[Function]}
    onWheel={[Function]}
    value="00:00"
  />
</div>
`;
